lib/theory/src/Theory/Constraint/System/JSON.hs

```diff
@@ -0,0 +1,394 @@
+    JSONGraphNodeMetadata { jgnPrems = map (factToJSONGraphNodeFact pretty "p:" n) 
+                                       $ zip [0..] $ L.get rPrems ru
+                          , jgnActs  = map (itemToJSONGraphNodeFact pretty "action") $ L.get rActs ru 
+                          , jgnConcs = map (factToJSONGraphNodeFact pretty "c:" n) 
+                                       $ zip [0..] $ L.get rConcs ru
+                          }
+
+-- generate JSONGraphNode from node of sequent
+nodeToJSONGraphNode :: Bool -> (NodeId, RuleACInst) -> JSONGraphNode
+nodeToJSONGraphNode pretty (n, ru) = 
+    JSONGraphNode { jgnId = show n
+                  , jgnType = getRuleType ru
+                  , jgnLabel = getRuleName ru
+                  , jgnMetadata = Just (nodeToJSONGraphNodeMetadata pretty (n, ru))
+                  }
+
+-- determine the type of an edge
+getRelationType :: NodeConc -> NodePrem -> System -> String
+getRelationType src tgt se =
+    let check p = maybe False p (resolveNodePremFact tgt se) ||
+                  maybe False p (resolveNodeConcFact src se)
+        relationType | check isKFact          = "KFact"
+                     | check isPersistentFact = "PersistentFact"
+                     | check isProtoFact      = "ProtoFact"
+                     | otherwise              = "default"
+    in
+    relationType
+
+-- generate JSON data structure for lastAtom
+lastAtomToJSONGraphNode :: Maybe NodeId -> [JSONGraphNode]
+lastAtomToJSONGraphNode n = case n of
+    Nothing -> [] 
+    Just n -> [JSONGraphNode { jgnId = show n
+                             , jgnType = "lastAtom"
+                             , jgnLabel = show n
+                             , jgnMetadata = Nothing 
+                             }] 
+
+-- generate JSON data structure for unsolvedActionAtom
+unsolvedActionAtomsToJSONGraphNode :: Bool -> (NodeId, LNFact) -> JSONGraphNode
+unsolvedActionAtomsToJSONGraphNode pretty (n, f) =
+    JSONGraphNode { jgnId       = show n
+                  , jgnType     = "unsolvedActionAtom"
+                  , jgnLabel    = case pretty of
+                                    True  -> pps $ prettyLNFact f
+                                    False -> ""
+                  , jgnMetadata = 
+                      Just JSONGraphNodeMetadata { jgnPrems = []
+                                                 , jgnActs  = [itemToJSONGraphNodeFact pretty "action" f] 
+                                                 , jgnConcs = [] 
+                                                 }
+                  }    
+
+-- generate a JSONGraphNode for those nodes in sEdges that are not present in sNodes 
+-- might occur in the case distinctions shown in the GUI
+-- since a fact is missing, the id is encoded as jgnFactId, could also be done directly in jgnId
+missingNodesToJSONGraphNodes :: System -> [Edge] -> [JSONGraphNode]
+missingNodesToJSONGraphNodes _ [] = []
+missingNodesToJSONGraphNodes se ((Edge (sid, _) (tid, _)):el) 
```

Comment:
Just as a suggestion: would it be possible to reduce the indents by adding newlines and 2-space indents before the `{`es. See the [Elevence style-guide](https://github.com/meiersi/HaskellerZ/blob/master/meetups/20160128-A_primer_to_commercial_Haskell_programming/code-by-elevence/docs/hs-style-guide.md) for the reasoning.)

Thanks for your consideration.